datastruct/mpool.h

```diff
@@ -11,69 +13,114 @@
  * of allowing the memory to be reused by some other code.
  */
 
+struct mpool {
+	size_t stacklen;
+	size_t allocsize;
+	void ** allocs;
+	uint64_t nallocs;
+	uint64_t nempties;
+	int state;
+	void ** allocs_static;
+	void (*atexitfunc)(void);
+};
+
+static inline void
+mpool_atexit(struct mpool * M)
+{
+
+	while (M->stacklen)
+		free(M->allocs[--M->stacklen]);
+	if (M->allocs != M->allocs_static)
+		free(M->allocs);
+}
+
+static inline void *
+mpool_malloc(struct mpool * M, size_t len)
+{
+
+	M->nallocs++;
+	if (M->stacklen)
+		return (M->allocs[--(M->stacklen)]);
+	M->nempties++;
+	if (M->state == 0) {
+		atexit(M->atexitfunc);
+		M->state = 1;
+	}
+	return (malloc(len));
+}
+
+static inline void
+mpool_free(struct mpool * M, void * p)
+{
+	void * allocs_new;
+
+	if (p == NULL)
+		return;
+
+	if (M->stacklen < M->allocsize) {
+		M->allocs[M->stacklen++] = p;
+		return;
+	}
+
+	if (M->nempties	> (M->nallocs >> 8)) {
+		allocs_new = malloc(M->allocsize * 2 * sizeof(void *));
+		if (allocs_new) {
+			memcpy(allocs_new, M->allocs,
+			    M->allocsize * 2 * sizeof(void *));
```

Comment:
this looks like undefined behaviour.  We've allocated
    `M->allocsize * sizeof(void *)`
bytes, but you're copying `2*` that value.  Removing the `2*` stops the segfaults I see.  (I haven't yet empirically checked that the mpool content is correct after the adjusted `memcpy`, but I can't see how it wouldn't be)

Reply:
*cough*

Umm, yeah.  When we're expanding a buffer from holding one record to holding two, we need to copy the *one* existing record, not copy two records from the buffer which only has one record in it. :-/

I'm glad I asked you to write some tests for this!